internal/crypto/crypto.go

```diff
@@ -233,79 +235,127 @@ func (k *EncryptionKey) Valid() bool {
-		ciphertext = append(ciphertext, make([]byte, ext)...)
+	if !validNonce(nonce) {
+		panic("nonce is invalid")
 	}
 
-	iv := newIV()
-	copy(ciphertext, iv[:])
+	dst = dst[:cap(dst)]
+
+	// extend dst if necessary
+	if len(dst) < len(plaintext)+k.Overhead() {
+		ext := len(plaintext) + k.Overhead() - len(dst)
+		dst = append(dst, make([]byte, ext)...)
+	}
 
 	c, err := aes.NewCipher(k.EncryptionKey[:])
 	if err != nil {
 		panic(fmt.Sprintf("unable to create cipher: %v", err))
 	}
-	e := cipher.NewCTR(c, ciphertext[:ivSize])
-	e.XORKeyStream(ciphertext[ivSize:], plaintext)
+	e := cipher.NewCTR(c, nonce)
+	e.XORKeyStream(dst, plaintext)
 
-	// truncate to only cover iv and actual ciphertext
-	ciphertext = ciphertext[:ivSize+len(plaintext)]
+	// truncate to only cover the ciphertext
+	dst = dst[:len(plaintext)]
 
-	mac := poly1305MAC(ciphertext[ivSize:], ciphertext[:ivSize], &k.MACKey)
-	ciphertext = append(ciphertext, mac...)
+	mac := poly1305MAC(dst, nonce, &k.MACKey)
+	dst = append(dst, mac...)
 
-	return ciphertext, nil
+	return dst
 }
 
-// Decrypt verifies and decrypts the ciphertext. Ciphertext must be in the form
-// IV || Ciphertext || MAC. plaintext and ciphertext may point to (exactly) the
-// same slice.
-func (k *Key) Decrypt(plaintext []byte, ciphertextWithMac []byte) (int, error) {
+// Open decrypts and authenticates ciphertext, authenticates the
+// additional data and, if successful, appends the resulting plaintext
+// to dst, returning the updated slice. The nonce must be NonceSize()
+// bytes long and both it and the additional data must match the
+// value passed to Seal.
+//
+// The ciphertext and dst may alias exactly or not at all. To reuse
+// ciphertext's storage for the decrypted output, use ciphertext[:0] as dst.
+//
+// Even if the function fails, the contents of dst, up to its capacity,
+// may be overwritten.
+func (k *Key) Open(dst, nonce, ciphertext, additionalData []byte) ([]byte, error) {
 	if !k.Valid() {
-		return 0, errors.New("invalid key")
+		return nil, errors.New("invalid key")
+	}
+
+	// check parameters
+	if len(nonce) != ivSize {
+		panic("incorrect nonce length")
+	}
+
+	if !validNonce(nonce) {
+		return nil, errors.New("nonce is invalid")
 	}
 
 	// check for plausible length
-	if len(ciphertextWithMac) < Extension {
-		return 0, errors.Errorf("trying to decrypt invalid data: ciphertext too small")
+	if len(ciphertext) < k.Overhead() {
+		return nil, errors.Errorf("trying to decrypt invalid data: ciphertext too small")
 	}
 
+	dst = dst[:cap(dst)]
+	plaintextLength := len(ciphertext) - k.Overhead()
+
 	// check buffer length for plaintext
-	plaintextLength := len(ciphertextWithMac) - Extension
-	if len(plaintext) < plaintextLength {
-		return 0, errors.Errorf("plaintext buffer too small, %d < %d", len(plaintext), plaintextLength)
+	if len(dst) < plaintextLength {
+		return nil, errors.Errorf("plaintext buffer too small, %d < %d", len(dst), plaintextLength)
```

Comment:
Open is an append-like API. It should not require a preallocated buffer, dst can be nil.

Reply:
Ah, you're right.